core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocateUnassignedDecision.java

```diff
@@ -265,13 +262,116 @@ public String getAllocationId() {
+     * the cluster and the deciders are waiting to see if the node returns before allocating the replica
+     * elsewhere.  Only returns a meaningful positive value if {@link #getAllocationStatus()} returns
+     * {@link AllocationStatus#DELAYED_ALLOCATION}.
+     */
+    public long getTotalDelayInMillis() {
+        return totalDelayInMillis;
+    }
+
+    /**
+     * Returns the explanation behind the {@link #getDecision()} that is returned for this decision.
+     */
+    public String getExplanation() {
+        String explanation;
+        if (decision == Type.NO) {
+            assert allocationStatus != null : "if the decision is NO, it must have an AllocationStatus";
+            if (allocationStatus == AllocationStatus.FETCHING_SHARD_DATA) {
+                explanation = "cannot allocate because information about existing shard data is still being retrieved from " +
+                                  "some of the nodes";
+            } else if (allocationStatus == AllocationStatus.NO_VALID_SHARD_COPY) {
+                if (nodeDecisions != null && nodeDecisions.size() > 0) {
+                    explanation = "cannot allocate because all existing copies of the shard are unreadable";
+                } else {
+                    explanation = "cannot allocate because a previous copy of the shard existed, but could not be found";
+                }
+            } else if (allocationStatus == AllocationStatus.DELAYED_ALLOCATION) {
+                explanation = "cannot allocate because the cluster is waiting " + Long.toString(remainingDelayInMillis / 1000L) +
+                                  " seconds for the departed node holding a replica to rejoin";
+            } else {
+                assert allocationStatus == AllocationStatus.DECIDERS_NO;
+                if (reuseStore) {
+                    explanation = "cannot allocate because allocation is not permitted to any of the nodes that hold a valid shard copy";
+                } else {
+                    explanation = "cannot allocate because allocation is not permitted to any of the nodes";
+                }
+            }
+        } else if (decision == Type.YES) {
+            explanation = "can allocate the shard";
+        } else if (decision == Type.THROTTLE) {
+            explanation = "allocation temporarily throttled";
+        } else {
+            throw new IllegalStateException("unhandled decision [" + decision + "]");
+        }
+        return explanation;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.field("decision", decision.toString());
+        builder.field("explanation", getExplanation());
+        if (allocationStatus != null) {
+            builder.field("allocation_status", allocationStatus.value());
+        }
+        if (assignedNode != null) {
+            builder.startObject("assigned_node");
+            builder.field("id", assignedNode.getId());
```

Comment:
Do you mean a single field `assigned_node` with to toString() of the `DiscoveryNode`?  Or calling `DiscoveryNode#toXContent`?  I feel it may be too verbose and provide unnecessary information, especially when any of that extra information is easily obtained via `/_cat/nodes` or getting the cluster state.  But if you feel strongly that we should put the full object there, I don't mind.